launch/launcher.go

```diff
@@ -166,77 +117,94 @@ func supportsExecD(bp Buildpack) bool {
+	for _, bp := range l.Buildpacks {
+		var skip bool
+		for _, pred := range predicates {
+			skip = skip || !pred(bp) // skip unless all predicates return true
+		}
+		if skip {
+			continue
 		}
-	}
 
-	appendFilesInDir := func(path string) error {
-		fis, err := ioutil.ReadDir(path)
-		if os.IsNotExist(err) {
+		dir := filepath.Join(l.LayersDir, EscapeID(bp.ID))
+		if _, err := os.Stat(dir); os.IsNotExist(err) {
 			return nil
+		} else if err != nil {
+			return errors.Wrap(err, "find buildpack directory")
 		}
-		if err != nil {
-			return errors.Wrapf(err, "failed to list files in dir '%s'", path)
+		if err := fn(dir); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		for _, fi := range fis {
-			appendIfFile(filepath.Join(path, fi.Name()), fi)
-		}
-		return nil
+func (l *Launcher) doLayerRoot() action {
+	return func(path string) error {
+		return l.Env.AddRootDir(path)
 	}
+}
 
-	if err := l.eachBuildpack(func(path string) error {
-		absPath, err := filepath.Abs(path)
-		if err != nil {
+func (l *Launcher) doLayerEnvFiles(proc Process) action {
+	return func(path string) error {
+		if err := l.Env.AddEnvDir(filepath.Join(path, "env")); err != nil {
 			return err
 		}
-		return eachDir(absPath, func(path string) error {
-			if err := appendFilesInDir(filepath.Join(path, dirName)); err != nil {
-				return err
-			}
-			if process.Type != "" {
-				return appendFilesInDir(filepath.Join(path, dirName, process.Type))
-			}
+		if err := l.Env.AddEnvDir(filepath.Join(path, "env.launch")); err != nil {
+			return err
+		}
+		if proc.Type == "" {
 			return nil
-		})
-	}, filters...); err != nil {
-		return nil, err
+		}
+		return l.Env.AddEnvDir(filepath.Join(path, "env.launch", proc.Type))
 	}
-	return files, nil
 }
 
-func (l *Launcher) eachBuildpack(fn func(path string) error, filters ...func(bp Buildpack) bool) error {
-Loop:
-	for _, bp := range l.Buildpacks {
-		for _, filter := range filters {
-			if !filter(bp) {
-				continue Loop
-			}
-		}
-		if err := fn(filepath.Join(l.LayersDir, EscapeID(bp.ID))); err != nil {
+func (l *Launcher) doLayerExecD(proc Process) action {
+	return func(path string) error {
+		if err := eachFile(filepath.Join(path, "exec.d"), func(path string) error {
+			return l.ExecD.ExecD(path, l.Env)
+		}); err != nil {
 			return err
 		}
+		if proc.Type == "" {
+			return nil
+		}
+		return eachFile(filepath.Join(path, "exec.d", proc.Type), func(path string) error {
+			return l.ExecD.ExecD(path, l.Env)
+		})
 	}
-	return nil
 }
 
-func eachDir(dir string, fn func(path string) error) error {
-	files, err := ioutil.ReadDir(dir)
+func eachLayer(bpDir string, action action) error {
+	return eachInDir(bpDir, action, func(fi os.FileInfo) bool {
+		return fi.IsDir()
+	})
+}
+
+func eachFile(dir string, action action) error {
+	return eachInDir(dir, action, func(fi os.FileInfo) bool {
+		return !fi.IsDir()
+	})
+}
+
+func eachInDir(dir string, action action, pred func(fi os.FileInfo) bool) error {
```

Comment:
```suggestion
func eachInDir(dir string, action action, predicate func(fi os.FileInfo) bool) error {
```
To be consistent with line 126.